### PR TITLE
Add new azure active directory related modules

### PR DESCRIPTION
**Pull request by Fred-sun**
##### ISSUE TYPE

- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request
- New Module Pull Request

##### COMPONENT NAME

Add new module.
##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: plugins/modules/azure_ad_password_info.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
+#!/usr/bin/python
+#
+# Copyright (c) 2020 Haiyuan Zhang, <haiyzhan@microsoft.com>
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+import datetime
+try:
+    from dateutil.relativedelta import relativedelta
```
Comment (haiyuazhang) [inline plugins/modules/azure_ad_password_info.py]: this import should be removed

File: plugins/modules/azure_ad_password_info.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+DOCUMENTATION = '''
+module: azure_ad_password_info
+
+version_added: "2.10"
+
+short_description: Get application info
+
+description:
+        - Get application info.
```
Comment (haiyuazhang) [inline plugins/modules/azure_ad_password_info.py]: Get application password info

File: plugins/modules/azure_ad_password_info.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+options:
+    app_id:
+        description:
+            - The application ID.
```
Comment (haiyuazhang) [inline plugins/modules/azure_ad_password_info.py]: Get Application password info

File: plugins/modules/azure_ad_password_info.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+      tenant: "{{ tenant_id }}"
+      key_id: "{{ key_id }}"
+'''
+
+RETURN = '''
+passwords:
+    description:
+        - Password info.
+    returned: success
+    type: complex
```
Comment (haiyuazhang) [inline plugins/modules/azure_ad_password_info.py]: complex

File: plugins/modules/azure_ad_password_info.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+            filtered = [pd for pd in passwords if pd.key_id == self.key_id]
+            self.results['passwords'] = [self.to_dict(pd) for pd in filtered]
+        else:
+            self.results['passwords'] = [self.to_dict(pd) for pd in passwords]
+
+        return self.results
+
+    def resolve_app_obj_id(self):
+        try:
+            if self.app_object_id is not None:
```
Comment (haiyuazhang) [inline plugins/modules/azure_ad_password_info.py]: app_object_id is not documented

File: plugins/modules/azure_ad_serviceprincipal.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+            to_update = {}
+            if self.app_role_assignment_required is not None:
+                to_update['app_role_assignment_required'] = self.app_role_assignment_required
+
+            client.service_principals.update(old_response['object_id'], to_update)
+            self.results['changed'] = True
+            self.results.update(self.get_resource())
+
+        except GraphErrorException as ge:
+            self.fail("Error deleting service principal app_id {0} - {1}".format(self.app_id, str(ge)))
```
Comment (haiyuazhang) [inline plugins/modules/azure_ad_serviceprincipal.py]: error updating the service principal

File: plugins/modules/azure_ad_serviceprincipal.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+        except GraphErrorException as ge:
+            self.log("Did not find the graph instance instance {0} - {1}".format(self.app_id, str(ge)))
+            return False
+
+    def check_update(self, response):
+        app_assignment_changed = self.app_role_assignment_required is not None and \
+            self.app_role_assignment_required != response.get('app_role_assignment_required', None)
+        to_be_update = False or app_assignment_changed
+
+        return to_be_update
```
Comment (haiyuazhang) [inline plugins/modules/azure_ad_serviceprincipal.py]: return at 203

File: plugins/modules/azure_ad_password.py [hunk 8]
```diff
@@ -0,0 +1,289 @@
+#!/usr/bin/python
+#
+# Copyright (c) 2020 Haiyuan Zhang, <haiyzhan@microsoft.com>
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+import datetime
+try:
+    from dateutil.relativedelta import relativedelta
```
Comment (haiyuazhang) [inline plugins/modules/azure_ad_password.py]: move the import clause below

File: plugins/modules/azure_ad_password.py [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+        for pd in old_passwords:
+            if pd.key_id == self.key_id:
+                old_passwords.remove(pd)
+                break
+        try:
+            self.client.applications.patch(self.app_object_id, ApplicationUpdateParameters(password_credentials=old_passwords))
+            num_of_passwords_after_delete = len(self.get_all_passwords())
+            if num_of_passwords_after_delete != num_of_passwords_before_delete:
+                self.results['changed'] = True
+                self.results['num_of_passwords_before_delete'] = num_of_passwords_before_delete
```
Comment (haiyuazhang) [inline plugins/modules/azure_ad_password.py]: delete

File: plugins/modules/azure_ad_password.py [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+                self.results['num_of_passwords_after_delete'] = num_of_passwords_after_delete
```
Comment (haiyuazhang) [inline plugins/modules/azure_ad_password.py]: delete

File: plugins/modules/azure_ad_password_info.py (hunk 3 above)
Comment (Fred-sun) [reply plugins/modules/azure_ad_password_info.py]: This is the description of the 'app_id', should not need to change!！

File: plugins/modules/azure_ad_password_info.py (hunk 5 above)
Comment (Fred-sun) [reply plugins/modules/azure_ad_password_info.py]: > app_object_id is not documented

59 to 62 lines

File: plugins/modules/azure_ad_password_info.py (hunk 1 above)
Comment (Fred-sun) [reply plugins/modules/azure_ad_password_info.py]: fixes

File: plugins/modules/azure_ad_password_info.py (hunk 2 above)
Comment (Fred-sun) [reply plugins/modules/azure_ad_password_info.py]: fixes

File: plugins/modules/azure_ad_password.py [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+        description:
+            - The application ID.
+        type: str
+    service_principal_id:
+        description:
+            - The service principal ID.
+        type: str
+    key_id:
+        description:
+            - Password key ID.
```
Comment (haiyuazhang) [inline plugins/modules/azure_ad_password.py]: The password key ID.

File: plugins/modules/azure_ad_password.py [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+        type: str
+        required: True
+    end_date:
+        description:
+            - Date or datemtime after which credentials expire.
+            - Default value is one year after current time.
+        type: str
+    value:
+        description:
+            - Application password value.
```
Comment (haiyuazhang) [inline plugins/modules/azure_ad_password.py]: The ...

File: plugins/modules/azure_rm_adserviceprincipal.py [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+    sample: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
+app_role_assignment_required:
+    description:
+        - Whether the Role of the Service Principal is set.
+    returned: always
+    type: bool
+    sample: false
+object_id:
+    description:
+        - Object ID of the associated application.
```
Comment (haiyuazhang) [inline plugins/modules/azure_rm_adserviceprincipal.py]: no sure if it is the object id of the associated service principal

**Pull request merged**